clients/src/main/java/org/apache/kafka/common/requests/ProduceResponse.java

```diff
@@ -67,20 +69,31 @@ public ProduceResponse(ProduceResponseData produceResponseData) {
      */
     @Deprecated
     public ProduceResponse(Map<TopicPartition, PartitionResponse> responses) {
-        this(responses, DEFAULT_THROTTLE_TIME);
+        this(responses, DEFAULT_THROTTLE_TIME, Collections.emptyList());
     }
 
     /**
-     * Constructor for the latest version
+     * Constructor for versions <= 9
      * @param responses Produced data grouped by topic-partition
      * @param throttleTimeMs Time in milliseconds the response was throttled
      */
     @Deprecated
     public ProduceResponse(Map<TopicPartition, PartitionResponse> responses, int throttleTimeMs) {
-        this(toData(responses, throttleTimeMs));
+        this(toData(responses, throttleTimeMs, Collections.emptyList()));
+    }
+
+    /**
+     * Constructor for the latest version
+     * @param responses Produced data grouped by topic-partition
+     * @param throttleTimeMs Time in milliseconds the response was throttled
+     * @param nodeEndpoints List of node endpoints
+     */
+    @Deprecated
```

Comment:
I'm confused—why is the new constructor is marked as `@Deprecated`? If that's intentional, can you add a comment about what should be used instead? Thanks.

Reply:
I didn't look too deep into this before, but this was deprecated in https://issues.apache.org/jira/browse/KAFKA-9628 and the follow up is described here https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/KafkaApis.scala#L605-L608. I'm not sure how big of a change that would be, and it is a critical part of the code, so I suspect it would be better to separate it into a different PR. I'm also not sure how noticeable the performance benefit will be. For now I think adding to this comment should be enough.

Reply:
I think all we would need to do is create the produceResponseData object. I don't think it would be too much work. I'm a little wary of adding more deprecated constructors.

Reply:
Taking a closer look, it looks like there was an effort to build the response directly and not pass in data structures (new maps) that will just be converted via toData.

Reply:
I started implementing this but I do think it's getting out of scope of this PR. [These](https://github.com/chb2ab/kafka/commit/e7bbd7b463db115e2f7b14618ea14da0ee893ae4) are my initial changes, to finish I think we want to remove PartitionResponse completely and replace it with PartitionProduceResponse. Having a deprecated constructor isn't ideal but I think we should remove it with KAFKA-10730, not this. @jolshan what do you think?

Reply:
KAFKA-10730 is a pretty dormant JIRA. I do agree that there is some level of conversion. I wonder if folks have a strong opinion about this conversion still. 

Looking into this further, I see the change would need to be made to appendRecords and the ProducePartitionStatus. It doesn't look too crazy, but also understandable this is not the scope for this PR.

I wonder if KAFKA-9682 was premature in deprecating the constructor. I guess our options are leaving it deprecated and adding a deprecated method or removing the deprecation until KAFKA-10730 is completed. (I almost just want to fix it so this doesn't happen in the future 😂 )

Reply:
I'll try to find time to finish the changes for [KAFKA-10730](https://issues.apache.org/jira/browse/KAFKA-10730), I think refactoring the tests would take some time but overall I agree it doesn't seem too big.

I'm ok with removing the deprecation, but I suspect the incentive to do the refactoring will be lost, so leaving it for now.

Reply:
@chb2ab It makes sense to do KAFKA-10730. But i agree with @jolshan that it is outside of the scope of the PR, should be done independently.

How about merging the PR with ctor as `@deprecated`? And then do follow-up PR for KAFKA-10730(tackling the new ctor as well).